### DISCUSSION TITLE
Howto revoke permission to open links in an app

### DISCUSSION THREAD

**Discussion by exaiwitmx** [MicrosoftEdge/WebView2Feedback#4479] [Q&A] [answered]
If you have granted the permission to always open specific links in an app, how do you revoke this permission?

![image](https://github.com/MicrosoftEdge/WebView2Feedback/assets/62023106/53a4c541-ec7b-4827-aa66-5532f2d4e88d)

What I've already tried:

1. WebView2.CoreWebView2.Profile.GetNonDefaultPermissionSettingsAsync()
=> No effect.

2. WebView2.CoreWebView2.Profile.GetNonDefaultPermissionSettingsAsync() with all CoreWebView2BrowsingDataKinds enumeration values OR-ed.
=> No effect.

3. WebView2.CoreWebView2.Profile.GetNonDefaultPermissionSettingsAsync()
=> Returns nothing.

4. Delete User Data Folder recursively by means of System.IO.Directory.Delete.
=> Actually removes the permission. But I don't want to do that for several reasons - including the problem that part of the User Data Folder might be locked (e.g. by other instances of my application) and therefore cannot be deleted.

I use version 1.0.2420.47 of the Microsoft.Web.WebView2 nuget package:
![image](https://github.com/MicrosoftEdge/WebView2Feedback/assets/62023106/6ad3e53f-c498-478b-ae32-89ff5ccd840b)

**Comment by victorhuangwq** [comment]
@vickiez do you have any idea about this?

**Reply by vickiez** [reply]
This looks like an external protocol scenario, which is handled by a separate API. + @maurawinstanley

**Reply by exaiwitmx** [reply]
@vickiez

> This looks like an external protocol scenario,

Yes, I can prevent the page from opening in the external application:

```
private void CoreWebView2_FrameNavigationStarting(object sender, CoreWebView2NavigationStartingEventArgs e)
{
	if (e.Uri.ToLowerInvariant().StartsWith("msteams:/"))
		e.Cancel = true;
}
```

> which is handled by a separate API.

I've found #137, but couldn't find any information on how to revoke the grant to always open such links in the external application.

**Comment by exaiwitmx** [comment]
Further information:

- The CoreWebView2.PermissionRequested event is not raised - not even if you haven't granted the permission before.

- In standalone MS Edge, the grant is revoked when clearing "Cookies and other site data":
![image](https://github.com/MicrosoftEdge/WebView2Feedback/assets/62023106/64cea752-afed-44a4-92d4-3153fc91f69b)

**Comment by exaiwitmx** [comment]
Is there any news about this?

**Comment by maurawinstanley** [comment, marked as answer]
@exaiwitmx Thanks for reaching out. It looks like we have not included that permission type in our [ClearBrowsingData API](https://learn.microsoft.com/en-us/microsoft-edge/webview2/concepts/clear-browsing-data?tabs=csharp). We will add an option to clear the external protocol data permissions and will update the thread when you can see the change in our Runtime. From there you can then use the Clear Browsing Data API to clear the permissions. In the meantime, although it is not ideal, please use the clearing the UDF workaround.

**Reply by vickiez** [reply]
@exaiwitmx  we'll post further updates in the issue - #4549. Thanks!
